apps/arweave/src/ar_wallets.erl

```diff
@@ -170,28 +170,38 @@ handle_call({add_wallets, RootHash, Wallets, Height, Denomination}, _From, DAG)
 handle_call({set_current, RootHash, Height, PruneDepth}, _, DAG) ->
 	{reply, ok, set_current(DAG, RootHash, Height, PruneDepth)}.
 
-handle_cast({init, Blocks, [{from_state, SearchDepth}]}, _) ->
-	?LOG_DEBUG([{event, init_from_state}, {block_count, length(Blocks)}]),
+handle_cast({init, Blocks, Args}, _) ->
+	SearchDepth = maps:get(search_depth, Args, length(Blocks)),
+	?LOG_DEBUG([{event, reading_account_tree_from_storage},
+			{block_count, length(Blocks)},
+			{search_depth, SearchDepth}]),
 	case find_local_account_tree(Blocks, SearchDepth) of
 		not_found ->
-			ar:console("~n~n\tThe local state is missing an account tree, consider joining "
-					"the network via the trusted peers.~n"),
-			timer:sleep(1000),
-			erlang:halt();
+			B =
+				case length(Blocks) >= ?STORE_BLOCKS_BEHIND_CURRENT of
+					true ->
+						lists:nth(?STORE_BLOCKS_BEHIND_CURRENT, Blocks);
+					false ->
+						lists:last(Blocks)
+				end,
+			case maps:get(trusted_peers, Args, []) of
+				[] ->
```

Comment:
With this new logic we'll:
1. always check local state first
2. only if it's not found locally *and* trusted peers are set we'll try to download from them

is that right?

Reply:
This change alone might have a pretty huge impact on the network by cutting down the traffic to the trusted peers + speeding up node launch time